_posts/en/newsletters/2019-05-14-newsletter.md

```diff
@@ -0,0 +1,648 @@
+maximum of around four billion scripts.  However, only one of those
+scripts would appear in any spending transaction and only 32 hashes
+directly related to the merkle tree would need to be generated to prove
+the script was part of the tree---this means that more complex scripts
+than our simple HTLC can gain much larger efficiency improvements than
+seen above (see an article focused on [MAST][] for more information).
+
+### Slight changes to scripting with Tapscript
+
+When using Bitcoin Script with Taproot, some of the rules have been
+changed, most notably:
+
+- **Unused and invalid opcodes redefined:** the bytes of opcodes that
+  were never assigned, or which were made invalid by Satoshi Nakamoto,
+  or which were created for upgrades and have not been used yet have all
+  been changed to have the semantics of an `OP_SUCCESS` opcode that
+  unconditionally renders any script containing that code valid.
+  Because future soft forks can only add new rules by making
+  previously-valid things invalid, this maximal validity
+  today will allow maximal flexibility in future soft forks.  Of course,
+  receivers get to choose what scripts they use to receive payments, so no
+  one should include an `OP_SUCCESS` opcode in
+  their scripts until a soft fork has given it a new consensus-enforced
+  meaning.
+
+- **Schnorr signatures required:** ECDSA signatures will not be accepted
+  in Tapscript.
+
+- **New script-based multisig semantics:** the current
+  `OP_CHECKMULTISIG` and `OP_CHECKMULTISIGVERIFY` opcodes are not
+  available in Tapscript.  There are two alternatives.  First, the
+  existing single-sig `OP_CHECKSIG` and `OP_CHECKSIGVERIFY` may be used
+  in series.  For example (2-of-2 multisig):
+
+      <A pubkey> OP_CHECKSIGVERIFY <B pubkey> OP_CHECKSIG
+
+    Second, a new `OP_CHECKSIGADD` (`OP_CSADD`) opcode may be used to
+    increment a counter if a signature matches a specified public key.
+    For example (2-of-3):
+
+      <A pubkey> OP_CHECKSIG <B pubkey> OP_CSADD <C pubkey> OP_CSADD OP_2 OP_EQUAL
+
+    This change is made to allow for batch verification of multiple
+    signatures, which can [significantly speed up
+    verification][bip-schnorr] compared to checking each signature
+    independently.
+
+- **Redefined sigops limit:** because verifying signatures is the most
+  CPU expensive operation in Bitcoin Script, an early version of Bitcoin
+  added a limit on the maximum number of signature-checking operations
+  (sigops) a block could contain, and versions of this limit were
+  applied to later P2SH and segwit v0 scripts.  However, one problem
+  with this limit is that miners must consider two factors when
+  trying to select what valid transactions to include in a block:
+  fee density (fee/vbyte) and sigop density (sigops/vbyte).  This is
+  much more difficult than just optimizing block composition based on a
+  single parameter, such as fee density (feerate).
```

Comment:
I don't like that we have "fee density (fee/vbyte)" in the previous sentence and "fee density (feerate)" here. Suggested change: "fee density (fee/vbyte) and sigop density (sigops/vbyte).  This is much more difficult than optimizing block composition based on just fee density."